internal/auth/client_kerberos.go

```diff
@@ -0,0 +1,477 @@
+
+		hostAddr := types.HostAddressFromNetIP(k.remoteAddr)
+
+		ok, err := a.Verify(k.client.keytab, k.client.config.ClockSkew, hostAddr, k.client.acceptor)
+		if err != nil {
+			return nil, "", false, message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Failed to verify the AP_REQ packet (is the acceptor correct?)",
+			)
+		}
+		if !ok {
+			return nil, "", false, message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Couldn't verify AP_REQ packet",
+			)
+		}
+
+		k.key = a.Authenticator.SubKey
+
+		ticket := a.Ticket
+		rep, err := krbmsg.NewAPRep(ticket, a.Authenticator)
+		if err != nil {
+			return nil, "", false, message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Failed to generate an AP_REP packet",
+			)
+		}
+
+		repToken := spnego.NewKRB5TokenAPREP()
+		repToken.APRep = rep
+		mar2, err:= repToken.Marshal()
+		asn1tools.AddASNAppTag(mar2, 060)
+		if err != nil {
+			return nil, "", false, message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Failed to marshal the AP_REP packet",
+			)
+		}
+
+		authCred, err := a.Authenticator.GetCredDelegation()
+		if err != nil {
+			k.client.logger.Info(message.Wrap(
+				err,
+				message.EAuthKerberosVerificationFailed,
+				"Failed to unmarshal the Authenticator delegation packet",
+			))
+			// Accept but no cred delegation
+			return mar2, k.username, false, nil
+		}
+
+		if authCred != nil && authCred.HasDelegation() {
+			var cred krbmsg.KRBCred
+			err = cred.Unmarshal(authCred.Deleg)
+			if err != nil {
+				k.client.logger.Info(message.Wrap(
+					err,
+					message.EAuthKerberosVerificationFailed,
+					"Failed to marshal the KRB_CRED packet",
+				))
+				return mar2, k.username, false, nil
+			}
+			err = cred.DecryptEncPart(ticket.DecryptedEncPart.Key)
+			if err != nil {
+				return nil, "", false, err
+			}
+
+			cacheCreds, err := cred.ToCredentials()
+			if err != nil {
+				return nil, "", false, err
+			}
+			cache := credentials.CCacheFromCredentials(cacheCreds)
+			mar, err := cache.Marshal()
+			if err != nil {
+				return nil, "", false, err
+			}
+			k.credentials = mar
+		}
+
+		return mar2, k.username, false, nil
+	}
+
+	return nil, "", false, fmt.Errorf("Invalid token")
+}
+
+const (
+	SSH_MSG_USERAUTH_REQUEST = 50
+)
+
+// GSSAPIMicField is described in RFC4462 Section 3.5
+type GSSAPIMicField struct {
+	// SessionIdentifier is a random string identifying the ssh connection
+	// (different from containerSSHs identifier)
+	SessionIdentifier string
+	// Request is the action that's being requested (50: SSH_MSG_USERAUTH_REQUEST)
+	Request byte
+	// UserName is the username that the user requests to log in as
+	UserName string
+	// Service is the service being used ('ssh-connection')
+	Service string
+	// Method is the authentication method in use ('gssapi-with-mic')
+	Method string
+}
+
+func (mic *GSSAPIMicField) unmarshal(b []byte) error {
+	err := ssh.Unmarshal(b, mic)
```

Comment:
We are introducing a dependency on the unstable SSH API here. I would strongly suggest we move this into a wrapper function in the SSH package or come up with a better API that does not require dependence on the SSH package.